packages/alfa-table/src/table.ts

```diff
@@ -1,684 +1,854 @@
+    let groups: Array<Group> = [];
+    let table: Array<Array<Array<number>>> = [];
+
+    // In addition to the list of cells and the table grid, we also keep track
+    // of which columns and rows have data cells. This information is used when
+    // determining the scope of header cells.
+    const data = { x: new Set<number>(), y: new Set<number>() };
+
+    // 5
+    if (element.children().isEmpty()) {
+      return Table.of(element, cells, groups);
     }
 
-    public get width(): number {
-      return this._table.width;
+    // 6
+    // Nothing to do
+
+    // 7
+    let children = element.children().filter(isElement);
+
+    // 8
+    skip(not(hasName("colgroup", "thead", "tbody", "tfoot", "tr")));
+
+    // 9
+    while (current().some(hasName("colgroup"))) {
+      // 9.1
+      // As this step contains several substeps inlined in the algorithm, its
+      // substeps have been extracted into a function of their own.
+      processColumnGroup(current().get());
+
+      // 9.2
+      advance();
+
+      // 9.3
+      skip(not(hasName("colgroup", "thead", "tbody", "tfoot", "tr")));
+
+      // 9.4
+      // Nothing to do
     }
 
-    public get height(): number {
-      return this._table.height;
+    // 10
+    let yCurrent = 0;
+
+    // 11
+    let downwardGrowing: Array<[cell: Cell, index: number]> = [];
+
+    // Steps 12-18 are repeated for as long as there are children left in the
+    // table.
+    while (current().isSome()) {
+      // 12
+      skip(not(hasName("thead", "tbody", "tfoot", "tr")));
+
+      // 13
+      if (current().some(hasName("tr"))) {
+        processRow(current().get());
+        advance();
+        continue;
+      }
+
+      // 14
+      endRowGroup();
+
+      // 15
+      if (current().some(hasName("tfoot"))) {
+        footers.push(current().get());
+        advance();
+        continue;
+      }
+
+      // 16
+      processRowGroup(current().get());
+
+      // 17
+      advance();
+
+      // 18
+      // Nothing to do
     }
 
-    public get element(): Element {
-      return this._table.element;
+    // 19
+    footers.forEach(processRowGroup);
+
+    // 20
+    // Nothing to do
+
+    // 21
+    // Before returning the final table, we go through all cells and assign them
+    // scopes and headers, if any.
+    return Table.of(element, cells.map(assignScope).map(assignHeaders), groups);
+
+    /**
+     * Ensure that the given position is available in the table.
+     */
+    function ensure(x: number, y: number): void {
+      for (let i = table.length - 1; i < y; i++) {
+        table.push([]);
+      }
+
+      const row = table[y];
+
+      for (let i = row.length - 1; i < x; i++) {
+        row.push([]);
+      }
     }
 
-    public get colGroups(): Iterable<ColumnGroup> {
-      return this._table.columnGroups;
+    /**
+     * Get the cells at the given position.
+     */
+    function get(x: number, y: number): Array<number> {
+      ensure(x, y);
+      return table[y][x];
```

Comment:
The x/y inversion here (and other places) looks super error-prone.
I agree it's what we want, but this is the kind of 2-liner that triggers so many alarms when reading it. I would say we need a comment (either a detailled global one or a small one each time) to remind future selves that this is correct.

Reply:
Only in two places are the table contents actually accessed and those two places reside in functions defined right next to one another. I'll add a comment in both places explaining the "why" of it 👍